google/cloud/storage/bucket_test.cc

```diff
@@ -81,32 +97,28 @@ TEST(BucketTest, GetMetadataTooManyFailures) {
 #endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
-TEST(BucketTest, GetMetadataPermanentFailure) {
-  auto mock = std::make_shared<testing::MockClient>();
-  Client client{std::shared_ptr<internal::RawClient>(mock)};
-
+TEST_F(BucketTest, GetMetadataPermanentFailure) {
 #if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
   EXPECT_CALL(*mock, GetBucketMetadata(_))
       .WillOnce(Return(std::make_pair(PermanentError(), BucketMetadata{})));
-  EXPECT_THROW(client.GetBucketMetadata("foo-bar-baz"), std::runtime_error);
+  EXPECT_THROW(client->GetBucketMetadata("foo-bar-baz"), std::runtime_error);
```

Comment:
Can we assert more error detail, something like EXPECT_THAT(...).HasSubstr() in google3?
My concern with just asserting the exception is there could me multiple reasons that exception is thrown, and some other code (or test) flaw could mask a bug because you're getting the exception thrown regardless of the specific case you're trying to test.

Reply:
Done. My searches did not find a cleaner way to assert attributes of the exception, and I am not sure I particularly like how these assertions look.  Let me know what you think

Reply:
I do like the fact that these caught a problem in the implementation class though :grin:

Reply:
Can you point me to where you made that change? I don't see any new HasSubstr calls in your test code, beyond the  ones you had there in a different part of the tests before.

Glad this helped uncover a problem.

Reply:
https://github.com/GoogleCloudPlatform/google-cloud-cpp/pull/787/commits/047ab927da465388a6203651593ac9abd754ea60#diff-4a22245009b28cd64a5aa074892c185aR88

would be the current place to look.  Basically one has to change:

```C++
EXPECT_THROW(expr, exception_type);
```

to:

```C++
EXPECT_THROW(try {
  expr;
} catch(exception_type const& ex) {
  // check stuff about ex here.
  throw; // throw again to check the type
}, exception_type);
```

and then clang-format(1) changes the code to look a little less readable, meh.

Reply:
Thanks